src/session/content/message_row/document/status_indicator.rs

```diff
@@ -0,0 +1,115 @@
+use gtk::gsk;
+use gtk::CompositeTemplate;
+
+use super::file_status::FileStatus;
+
+mod imp {
+    use std::cell::Cell;
+
+    use super::*;
+
+    #[derive(Debug, Default, CompositeTemplate, glib::Properties)]
+    #[template(string = r#"
+        using Adw 1;
+
+        template $MessageDocumentStatusIndicator : Adw.Bin {
+            child: Image status_image {};
+        }
+    "#)]
+    #[properties(wrapper_type = super::StatusIndicator)]
+    pub(crate) struct StatusIndicator {
+        #[property(get, set = Self::set_masked)]
+        pub(super) masked: Cell<bool>,
+        #[template_child]
+        pub(super) status_image: TemplateChild<gtk::Image>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for StatusIndicator {
+        const NAME: &'static str = "MessageDocumentStatusIndicator";
+        type Type = super::StatusIndicator;
+        type ParentType = adw::Bin;
+
+        fn class_init(klass: &mut Self::Class) {
+            klass.set_css_name("indicator");
+            klass.bind_template();
+        }
+
+        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for StatusIndicator {
+        fn constructed(&self) {
+            self.parent_constructed();
+            self.obj().set_overflow(gtk::Overflow::Hidden);
+        }
+
+        fn properties() -> &'static [glib::ParamSpec] {
+            Self::derived_properties()
+        }
+
+        fn set_property(&self, id: usize, value: &glib::Value, pspec: &glib::ParamSpec) {
+            self.derived_set_property(id, value, pspec)
+        }
+
+        fn property(&self, id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+            self.derived_property(id, pspec)
+        }
+    }
+
+    impl WidgetImpl for StatusIndicator {
+        fn snapshot(&self, snapshot: &gtk::Snapshot) {
+            if self.masked.get() {
+                let widget = self.obj();
+
+                let color = widget.color();
+                let width = widget.width() as f32;
+                let height = widget.height() as f32;
+
+                snapshot.push_mask(gsk::MaskMode::InvertedAlpha);
+
+                self.parent_snapshot(snapshot);
+
+                snapshot.pop();
+
+                snapshot.append_color(&color, &gtk::graphene::Rect::new(0.0, 0.0, width, height));
+
+                snapshot.pop();
+            } else {
+                self.parent_snapshot(snapshot);
+            }
+        }
+    }
+
+    impl BinImpl for StatusIndicator {}
+
+    impl StatusIndicator {
+        fn set_masked(&self, masked: bool) {
+            if self.masked.replace(masked) != masked {
+                self.obj().queue_draw();
+            }
+        }
+    }
+}
+
+glib::wrapper! {
+    pub(crate) struct StatusIndicator(ObjectSubclass<imp::StatusIndicator>)
+        @extends adw::Bin, gtk::Widget;
+}
+
+impl StatusIndicator {
+    pub(crate) fn set_status(&self, status: &FileStatus) {
```

Comment:
I think, a reference isn't needed in this case as `FileStatus` only contains small data.

```suggestion
    pub(crate) fn set_status(&self, status: FileStatus) {
```

Reply:
Currently it's size is 16 bytes, twice the size of a reference)

Reply:
I'll derive Copy for it and pass everywhere by value

Reply:
Deriving Copy is ok, or you could just move it into the function. For such small sized data types, it is not significant in the end. What's more important that removing the `&` makes the code prettier.